### PR TITLE
MRG: FAQ update

### LINKED ISSUE #2944: Q: Get chan types?
Since we can't pickle evoked data (https://github.com/mne-tools/mne-python/issues/2899, again today, big PITA to read old data), and it is sometimes not reasonable to separately store MEG data from other variables (e.g. stats, conditions etc), I was thinking of keeping the necessary MEG `data` and `info` to recreate an `EvokedArray` for later plotting.

I think most params are easy to get (`ch_names`, `tmin`, `sfreq`) but `ch_types`, which, from what I've seen, needs to be created manually (there isn't an `evoked.info['ch_types']`).

If I'm not mistaken, could we add such functionality, or this would create some API issues? Or better, is there an easier way to approach this kind of issue?

### LINKED ISSUE #2945: DOC: better persistence doc and discourage pickling
cc @kingjr

### LINKED ISSUE #3002: Computing morph matrix on the fly is significantly faster than morph_data
Using `mne.compute_morph_matrix` and `mne.morph_data_precomputed` is ~100x faster than `mne.morph_data`.

The relevant morph map already exists and isn't re-written by either command.

**Output:**
```
Data to be morphed:
<SourceEstimate  |  5124 vertices, tmin : -200.0 (ms), tmax : 1200.0 (ms), tstep : 1.0 (ms), data size : 5124 x 1401> 
__________________________________________________
Compute morph matrix and morph_data_precomputed: 0.895 s
Morph_data without precomputed morph matrix: 103.332 s
```

**Script:**
```
stc = mne.read_source_estimate('processed_files/stcs/factorial/no_morph/noun/A0008_noun_no_morph-lh.stc')
fs = mne.read_source_spaces('mri/fsaverage/bem/fsaverage-ico-4-src.fif')
vertices_to = [fs[0]['vertno'], fs[1]['vertno']]

print "Data to be morphed:\n", stc, '\n', 50*'_'


t0 = time.time()
morph_mat = mne.compute_morph_matrix('A0008', 'fsaverage', vertices_from=stc.vertices, vertices_to=vertices_to, subjects_dir='mri')
stc_morphed = mne.morph_data_precomputed('A0008', 'fsaverage', stc, vertices_to, morph_mat)
delta_t = time.time()-t0

print "Compute morph matrix and morph_data_precomputed: %.3f s" %delta_t

t0 = time.time()
stc_morphed = mne.morph_data('A0008', 'fsaverage', stc, grade=vertices_to, subjects_dir='mri')
delta_t = time.time()-t0

print "Morph_data without precomputed morph matrix: %.3f s" %delta_t
```

### LINKED ISSUE #3009: isinstance('ch_name', str) should be checked against basestring
Checking against str fails for unicode channel names. Specifically, this problem occurs here
https://github.com/mne-tools/mne-python/blame/master/mne/preprocessing/eog.py#L105
for ch_name=u'EXG1'

Possibly, this is not the only place where this is problematic.

### LINKED ISSUE #3021: ADD to FAQ resample versus decimate methods
I have a question about the resample method for `mne.source_estimate.SourceEstimate`. why do we have resample here and how is it different from the `decimate` method that we have the `Epochs`? Is it generally safe to resample the `SourceEstimate` if it is doing internal filtering?

### PR DESCRIPTION
**Pull request by Eric89GXL**
A bunch of touch-ups and additions to the FAQ to get us closer to 0.12.

I also updated our `conf.py` to look like the upstream one from `sphinx-gallery` because otherwise I couldn't build the doc :(

Closes #3021.
Closes #3009.
Closes #3002.
Closes #2945.
Closes #2944.

### CODE REVIEW

File: doc/faq.rst [hunk 1]
```diff
@@ -61,6 +61,92 @@ reporting a bug, which should look something like this::
     skcuda:        Not found
 
 
+Why shouldn't I "pickle" my MNE-Python objects for later use?
```
Comment (jasmainak) [inline doc/faq.rst]: sounds like an odd way to frame the question ... how about: " I can't load my "pickled" object into MNE-Python after I upgraded." -- or something like that.

File: doc/faq.rst [hunk 2]
```diff
@@ -61,6 +61,92 @@ reporting a bug, which should look something like this::
...
+rate of data. We'll discuss some major ones here, with some of their
+implications:
+
+- :func:`mne.io.Raw.resample` is used to resample (typically downsample) raw
+  data. Resampling is the two-step process of applying a low-pass FIR filter
+  and subselecting samples from the data.
+
+  Using this function to resample data before forming :class:`Epochs`
+  for final analysis is generally discouraged because doing so effectively
+  loses precision (and jitters) event timings, see
```
Comment (jasmainak) [inline doc/faq.rst]: precise ?

**Comment by jasmainak** [comment]
The docs look like great additions. Thanks @Eric89GXL :)

File: doc/faq.rst (hunk 1 above)
Comment (Eric89GXL) [inline doc/faq.rst]: I'd rather immediately discourage with the question itself, though

File: doc/faq.rst (hunk 2 above)
Comment (Eric89GXL) [inline doc/faq.rst]: Ahh yeah misplaced parens

File: doc/faq.rst (hunk 1 above)
Comment (jasmainak) [inline doc/faq.rst]: ok, fine :)

File: doc/faq.rst [hunk 3]
```diff
@@ -61,6 +61,92 @@ reporting a bug, which should look something like this::
...
+  frequency, so this acts as a simplistic, non-ideal low-pass filter;
+  this will reduce but not eliminate aliasing if data were not sufficiently
+  low-passed. In the case where the "filter" or bin-width is a single sample
+  (i.e., an impulse) this operation simplifies to decimation without filtering.
+
+Resampling raw data is taking forever! What do I do?
+----------------------------------------------------
+
+:func:`mne.io.Raw.resample` was significantly sped up for version 0.12 by
+using the arameter ``npad=='auto'``. Try it, it might help!
```
Comment (kingjr) [inline doc/faq.rst]: Maybe ref to cuda here?

File: doc/faq.rst [hunk 4]
```diff
@@ -61,6 +61,92 @@ reporting a bug, which should look something like this::
...
+
+
 Inverse Solution
```
Comment (kingjr) [inline doc/faq.rst]: Shouldn't this following section (inverse modeling, covariance etc) below be in the tutorial or examples? My expectation is that the FAQ should relate to setup, i/o, memory, classic bugs as opposed to scientific questions.

Comment (Eric89GXL) [inline doc/faq.rst]: Let's talk about that over in the big DOC issue #1495 so we don't derail
here

Comment (kingjr) [inline doc/faq.rst]: ok

File: doc/faq.rst [hunk 5]
```diff
@@ -117,12 +203,38 @@ For expert use cases or debugging the alternative estimators can also be compare
...
+taking surface-based source space data from one subject and
+morphing it to another using a smoothing procedure. However, they can
+take different amounts of time to perform the computation.
+
+To use :func:`mne.SourceEstimate.morph_precomputed`, you must first
+precompute a morphing matrix with :func:`mne.compute_morph_matrix` which
+can take some time, but then the actual morphing operation carried out by
+:func:`mne.SourceEstimate.morph_precomputed` is very fast, even for
+:class:`mne.SourceEstimate` objects with many time points. The method
+:func:`mne.SourceEstimate.morph`, by contrast, smooths the data directly
```
Comment (teonbrooks) [inline doc/faq.rst]: smooths the data directly on the data?

File: doc/faq.rst [hunk 6]
```diff
@@ -117,12 +203,38 @@ For expert use cases or debugging the alternative estimators can also be compare
...
+on the data, which can be **very slow** with many time points.
+If there are thousands of time points, then
+:func:`mne.SourceEstimate.morph_precomputed` will be much faster; if there
+are a few time points, then :func:`mne.SourceEstiamte.morph` will be faster.
```
Comment (teonbrooks) [inline doc/faq.rst]: `mne.SourceEstiamte` --> `mne.SourceEstimate`

File: doc/index.rst [hunk 7]
```diff
@@ -122,7 +122,7 @@ Direct financial support for the project has been provided by:
 
    <h2>Community</h2>
 
-* `Analysis talk: join the MNE mailing list <http://mail.nmr.mgh.harvard.edu/mailman/listinfo/mne_analysis>`_
+* `Analysis talk: join the MNE mailing list <MNE mailing list>`_
 
 * `Feature requests and bug reports on GitHub <https://github.com/mne-tools/mne-python/issues/>`_
```
Comment (teonbrooks) [inline doc/index.rst]: replace with `GitHub issues page`

File: doc/faq.rst (hunk 3 above)
Comment (jaeilepp) [inline doc/faq.rst]: and aramater -> parameter

File: doc/faq.rst (hunk 1 above)
Comment (dengemann) [inline doc/faq.rst]: yes I agree with @Eric89GXL -- maybe rephrase: "why something atrocious is going to happen if you consider to use pickling !!!"

gif-keyboard-17370554860503284834

File: doc/faq.rst [hunk 8]
```diff
@@ -61,6 +61,92 @@ reporting a bug, which should look something like this::
...
 
+Why shouldn't I "pickle" my MNE-Python objects for later use?
+-------------------------------------------------------------
+`Pickling <https://docs.python.org/3/library/pickle.html>`_ data and
+MNE-Python objects for later use can be tempting due to its simplicity
+and generality, but it is usually not the best option:
+
+- http://www.benfrederickson.com/dont-pickle-your-data/
+- http://stackoverflow.com/questions/21752259/python-why-pickle
+
```
Comment (dengemann) [inline doc/faq.rst]: "short it is not made for stable persistence and it is very likely that you will not be able to read your data in a future not too far from the moment you saved them." or so.

File: doc/faq.rst [hunk 9]
```diff
@@ -61,6 +61,92 @@ reporting a bug, which should look something like this::
...
+  for final analysis is generally discouraged because doing so effectively
+  loses precision (and jitters) event timings, see
+  `this gist <https://gist.github.com/Eric89GXL/01642cb3789992fbca59>`_ as
+  a demonstration. However, resampling raw data can be useful for
+  (at least):
+
+    - Computing projectors in low- or band-passed data
+    - Exploring data
+
+- :func:`mne.preprocessing.ICA.fit` has a similar effect as
```
Comment (dengemann) [inline doc/faq.rst]: no this is not true, it happens of on a copy of the data.

Comment (dengemann) [inline doc/faq.rst]: it is only for fitting the model, nothing is changed.

File: doc/faq.rst [hunk 10]
```diff
@@ -117,12 +203,38 @@ For expert use cases or debugging the alternative estimators can also be compare
...
+:func:`mne.SourceEstimate.morph`, by contrast, smooths the data directly
+on the data, which can be **very slow** with many time points.
+If there are thousands of time points, then
+:func:`mne.SourceEstimate.morph_precomputed` will be much faster; if there
+are a few time points, then :func:`mne.SourceEstiamte.morph` will be faster.
+For data sizes in between, we advise testing to determine which is best,
+although some developers choose to always use
+:func:`mne.SourceEstimate.morph_precomputed` since it will rarely take
+a long time.
+
```
Comment (dengemann) [inline doc/faq.rst]: we could also point out the source space morphing here.

File: doc/faq.rst (hunk 1 above)
Comment (jona-sassenhagen) [inline doc/faq.rst]: Definitely include that picture.

File: doc/faq.rst (hunk 10 above)
Comment (Eric89GXL) [inline doc/faq.rst]: I'd rather leave that one to another PR once people use it and get a sense for the best practices. Maybe you already know enough to write it, but I've mostly forgotten :)

File: doc/index.rst (hunk 7 above)
Comment (Eric89GXL) [inline doc/index.rst]: I'd rather leave this one shorter, it's in the right sidebar

**Comment by Eric89GXL** [comment]
Comments addressed, thanks for the quick reviews

**Comment by Eric89GXL** [comment]
@agramfort did you tell CircleCI to run, or did it do that on its own?

**Comment by agramfort** [comment]
I did nothing

**Comment by agramfort** [comment]
LGTM

+1 for merge

thx @Eric89GXL

**Comment by dengemann** [comment]
Thanks @Eric89GXL --- great to have!
